### ISSUE TITLE
Suggestion: Alter Drop Chances

### ISSUE THREAD

**Issue by deilann** [HabitRPG/habitica#2479]
Now that Cole's fixed the drop issue, I've noticed something-- the amount of food you get isn't logical for how the system works.

It takes 1 hatching potion and 1 egg to make a pet, but 9 (correctly chosen) foods to evolve a pet. While food drops slightly more than eggs or potions, it's not weighted in a way that over time, won't leave a user with far more eggs and hatching potions than necessary to gain all mounts and pets.
##

**Comment by Jotempe** [comment]
The problem if you rebalance it is, that beginner players will be getting pets much more slowly, but stockpiling useless (for them) food.

**Comment by deilann** [comment]
Yeah, I thought about that. There could be some ways to get around that (altering algorithm based on level or number of pets), but I feel like the algorithm should be the same for all users. I'm thinking rather than 30/30/40 cutting it back to 20/20/60 would show a lot of results in the long run, but have a smaller impact in the short run?

**Comment by ghost** [comment]
But if you want all the pets AND all the mounts (rather than evolving all the pets into mounts leaving you with no pets) then you will end up with lots of useless food if the ratio is changed to favour food. Just a thought.

**Comment by deilann** [comment]
No it won't. In order to get ALL the pets and ALL the mounts, you would
need 180 eggs, 180 hatching potions, and at least 810 food.

90 pets fed to mounts = 810 food (at the very least, assuming it's all
preferred)
rehatching the pets, 90 more eggs, 90 more hatching potions.

**Comment by MarkVaughn** [comment]
My suggestion would be adjusting  the relationships to 25,25,50 vs 30,30,40.
As deilan mentioned right now you need 180 potions, 180 eggs and 810 food. 
1170 items total, 180 represents 15% of that yet the drop rate is 30%
The statistical perfect split would be 15/15/70, but i'm not suggesting that.
I agree that jumping from 30/30/40 to 20/20/60 could make a much more radical impact, that's why a 25/25/50 would be a good compromise.

**Comment by Gonazar** [comment]
Late to this party, but I totally agree. I've been playing for about 5 months and chose the rogue class, I already have 2/3rds of the pets and I'm stockpiled with eggs and potions that are useless to me. With this incredibly unbalanced drop rate I've only managed to get enough food drops to make 4 mounts.

Let's look at the probability: it's 60% chance to get items for hatching and only 40% to get items for making mounts. However, you only need 2 items for making pets vs 9 for making mounts. In 100 drops you could have 30 eggs, 30 pots and 40 food. Even with the type-probability added in, 12 of 30 pots would be commons, but you'd be able to mix and match with eggs. Assuming you had no pets to begin with, there's a very good chance after 100 drops, you'll have 30 pets (1/3rd of all pets just to remind you). Whereas the food, with the type-probability added (1-in-10), you'd have at best enough food to make 4 mounts, or at worst 4/9 pieces towards making 10 mounts.

If this continued on the best case scenario where you got lucky enough to get exactly the food you need, then it'd take 2250 drops to get 90 mounts, but in that time you would have also collected enough potions and eggs to make 675 pets. In the worst case scenario, you'd need to do this 10 times as much assuming you're really unlucky and get the same food over and over again. At the median point, needing 5 times as many drops in order to get the right food for 90 mounts, you'd still need to collect 11250 drops which is enough for 6750 pets. I think it's very clear that the drop rate is very unbalanced. 

Like Mark mentioned, the statistical split would be 15/15/70, but there is also the sub-probability of what type of drop you will get (base/skelleton/golden etc). Even with the best drop rate split you're still gonna end up hording a bunch of items that you don't really need. Why compound this with a drop ratio that makes it very unlikely to collect the items you need?

It gets even more complicated when you add in the drop-cap (DC) to the 2250 figure. It could take anywhere between 2 years for subscribers with 100+ perception (DC = 18) and 3.5 years for non-subscribers (DC = 9) for BEST CASE. After 5 months I'm lvl 31 with 64 PER with all my bonuses, so my drop cap is only 7. Not everyone is a rogue either. Looking through the Hall, the best perception I've seen on a non-rogue player was a lvl 70 healer at 90 PER, but otherwise most warriors and healers may not get a drop-cap higher than 7 or 8 til they're very high level.

> The problem if you rebalance it is, that beginner players will be getting pets much more slowly, but stockpiling useless (for them) food.

This is not a problem at all, in fact it's ideal because beginners will be able to stockpile food for when they get the pet they want. Having pets/eggs/potions isn't useful either, there are no benefits to having a pet other than appearing in your avatar so it doesn't matter if you're stockpiling eggs, potions or food at the beginning.

When you look at the collection POV of Habit, mounts is end game and pets should be a stepping stone roughly mid way. At the moment it's not exactly a challenge to collect all the pets while eggs and potions are dropping so frequently. If we look at my experience, I think I'll probably have all the pets after 8 months of playing (maybe a bit more to get all those rare hatching pots, but I'm already 3/9 of the golden and 15/27 on the rares). However at my current rate I think it'll take a couple years to get all the mounts which is very frustrating. 

Ideally, if end game takes, for example, 4 years and it takes 2 stages to reach mounts, then I'd expect it take 2 years to get all the pets and HALF the mounts. There would be a much more linear progression of collecting pets and making mounts. And those last 2 years, you'd get to collect all the pets again since the first generation would be all made into mounts.

/frustrated rant

PLEASE fix the drop rates.

**Comment by SabreCat** [comment]
I could totally see amping up the food drops as suggested here. Or maybe having a single food drop "hit" throw two pieces of random food instead of one.

**Comment by m1rrag** [comment]
tl;dr to Gonazar but only because I think you're saying what I wanted to say- I chose rogue so I could get more food drops and actually get a mount at all. 
- Currenlty I'm a level 23 Rogue with only one steed. (22 pets total)
- I have **551 gold coins but the only things I could buy with them are a hood, short sword, or health potion, all of which I have better versions of or don't need**
- I also have 12 eggs, one piece of candy, and zero hatching potions.
- Oh, and my gross perception is 81.5

I don't think I've ever gotten food from using pickpocket, which seems silly. Anyways please read what Gonazar said; I'm starting to lose interest in the app as a whole because there is no game in just gaining a bunch of gold you can't even spend on anything/ not being able to move up in a meaningful way.

**Comment by Jotempe** [comment]
Look boys, the basic idea is that you don't get the thing too fast, because you lose interest then. I'm playing since uhm, almost two years? And I finally collected full set of "basic" mounts like a few weeks ago. Now I'm in the process of collecting quest mounts. In order to keep people interested, the game needs both "short range" and "long range" goals. Getting pets is a shorter range, getting mounts a longer range goal. If you rebalance as you suggest, getting pets would become more difficult, but getting mounts would be easier, moving those goals closer to each other. Beginner gamers would complain that it is so difficult to get pets, while they have lots of useless food, while seasoned players would have less goals to aspire to, so they'll lose interest.

To m1rrag: the idea of gold is that you <b>yoursef</b> create opportunities of spending it - the rewards. If you balance the cost of your rewards correctly to the tasks you set up, then gold will become an important resource. So, put a reasonable price on those donuts you like to eat, those other computer games you like to play, on every half hour of watching TV, and on all other things you like to do, but feel, that you should control your use of.

**Comment by Gonazar** [comment]
Short range goals would be equipment since those can be achieved in a relatively short amount of time.
Pets and Mounts **together** make up the long term goals. We're not saying you should be able to achieve long term in less time, 2 years as you've done should be fine. We're saying that it should be more linear, like it takes 1 year to get pets and still 2 years (or maybe 3) to get all mounts, but you should have at least half the mounts by half of the expected completion time.

It shouldn't make a difference to beginner users since built up food will be used eventually and they're only just learning what the collection rate is. Sure maybe the game is allowing new users to get pets quickly to maintain interest, but then they'll develop expectations for the collection rate to continue at that rate or only dip slightly later on. It's teaching new users to expect to be able to collect things quickly. 

Pets serve no purpose back to the rest of the game other than completing the collection and working up to mounts. If it takes such a short time to collect pets, then it's not a challenge. In fact the opposite is true, after collecting all the pets I would expect players to lose interest because collection becomes so much slower. In just a few months I completed all my armor sets and I'll soon have all the pets. But after that I'm basically going to lose the collection aspect of HabitRPG because food rarely drops and mounts trickle in once every 2 months. 

Additionally, the way it is now, the drops are even more useless than if we had built up food. Food you could save up and use eventually on your pets, but you really can't do anything with extra eggs and potions. After 10 base potions I will never need them again, or similarly after 9 lion eggs, I'll never use them again. However it'll take 90 meat before I don't need it. That can be saved until I need it, or for new users who don't expect to get a pet that can use it anytime soon, it could be sold.

For experienced players selling items is quite useless since they're not worth much to us, just 2 or 3 gold. Compared to how much gold you get from completing tasks/dailies, this is nothing.
Selling is only useful to new users who don't have enough tasks to make money or haven't yet been able to commit to their habits. But new users aren't going to sell pets if they're trying to collect them, they'll only sell what they have extra of.

**Comment by Alys** [comment]
As an experienced player, I definitely find selling eggs and potions to be useful. The trick with gold is to come up with a reward scheme that lets you spend gold at roughly the same rate you are earning it - and that can indeed be very tricky! It took me nearly two years to develop rewards that really worked for me. But once you have managed that, the little bit of extra gold you can get from eggs and potions can be a very nice bonus. If you're close to a reward that you really want, a few eggs can let you achieve it immediately. Or if you're using gold for fines, the eggs can replenish your gold supplies when you've had to fine yourself severely. Excess eggs and potions are also VERY useful after death or rebirth, because you don't lose them, whereas you do lose all your gold.

I can't really comment on the other points made here because the drop rates are working well for me. I can see though that others could certainly have different opinions, and I wouldn't object to the rates being changed as long as it didn't swing too far towards food. It is hard to balance the game so that it works well for everyone. :(  We are working on customisable difficulty settings, and that could perhaps be used to allow players to choose between a couple of drop balance options.

**Comment by Gonazar** [comment]
Well if that works for you that's great. Though it wouldn't make a
difference what items you were selling to achieve that.

I think the majority of new users may not see it this way. It highly
depends on their ability to come up with self rewards. Honestly, if you're
unable to use that part of the system, then the default HabitRPG rewards
won't last as incentives for very long.

I get that the idea is to eventually become independent of the system with
your own rewards, but if you're gamifying habits, then I would expect more
game based rewards.

If the drop system can't be changed then I'd suggest implementing at least
one more mid-level collection. Something that would be slower than getting
pets but isn't as painfully slow as collecting mounts.

Looking past the drop rate, the fundamental issue is that the long-term
in-game goals take far longer than the short term goals, providing few
incentives in between.

Cheers,
Doug

**Comment by Alys** [comment]
_" Though it wouldn't make a difference what items you were selling to achieve that."_
Eggs and potions gain you more gold than food.

_"I think the majority of new users may not see it this way. "_
That's very true but they are less likely to want to sell eggs and potions because they'll still be collecting the pets.

**Comment by m1rrag** [comment]
Yes. If real-life rewards I made up for myself actually worked, I never would have used habitRPG in the first place.

I'd LOVE it if I could gamify my life IRL, but my organizational issues (and lack of funds) (and lots of other factors) mean that complicating my life in that way will actually bring my bad habits back. A goal I've been good at maintaining has been to keep off weight, so a food reward doesn't make any sense. Denying myself food so I can get a food reward is what gave me the overeating habits I had in the first place. I can't buy myself things because I haven't been getting enough hours at work and though I'm trying to change this, it's still largely outside of my control. And even when I do reward myself on habitRPG for filling out applications or attending work or whatever, I just get more game money I can't use on anything. 

@Jotempe if you  can think up some self-made out-of-game rewards I can add to my rewards list without using up my own real-life or financial assets, or attention-span, or have a thread on the subject that you can direct me to, I would love that.

**Comment by Alys** [comment]
You might find some ideas here: [Sample Custom Rewards](http://habitrpg.wikia.com/wiki/Sample_Custom_Rewards)
although I know that at least some of them are the types of rewards that won't work for you.

**Comment by MaidM** [comment]
I'd like to wade in in support of the comments that spending gold on custom IRL rewards doesn't really work for me either, and thank you m1rrag for laying out the reasoning so eloquently.  

What about collectable pieces that you then have to go on a quest to fuse into cool armour, weapons, etc?  Give more equippable options, generally, and make today's buyable options more varied and more random so that you are encouraged to buy that rare thing that's turned up in the market today since it probably won't be there tomorrow.  Maybe sometimes pieces of a collectable series turn up as buyable for gold, but you're otherwise forced to quest to find them.  I'm rich as Croesus after a couple weeks playing, but nothing to spend my gold on :(

The reason, however, that I came to this thread in the first place was to ask whether someone's been tinkering with the drop code.  I am way way way heavy on eggs over food and hatching potions.  I'm picking up drops fairly quickly as I have a short term habit that I'm hitting several times a day to try and finish it off.  However, my current inventory is 29 eggs, no hatching potions, and 8 food, and if I'm estimating correctly, I've picked up a total of around 50-60 food, 27 hatching potions and 56 eggs.  Each day, I pick up several eggs, but if I'm lucky I might net 2 potions.  I've looked at the wiki version of the code, which seems to correctly 30-30-40 split the drop; is this still in play?  Or is the randomiser possibly at fault?  If it's randomising based on datetime stamps, it would make sense that recording a lot of habits at the one time would really skew the probabilities toward one type of drop.

**Comment by Alys** [comment]
@MaidM  We haven't been tinkering with the drop code and it doesn't randomise on datetime stamps. You can see the current code here: https://github.com/HabitRPG/habitrpg/blob/develop/common/script/index.coffee#L1251-1329
True randomness means that sometimes some people will get long streaks of one thing or another.

**Comment by MaidM** [comment]
@Alys But random number generators are notorious for not being truly random.  It's this line I'm querying the code on:
rarity = user.fns.predictableRandom(user.stats.gp)

What does predictableRandom do? Is it using user.stats.gp as a seed?  I appreciate that streaks happen with true randomness, but over large sets of results one should observe something approaching the intended distribution.  My anecdotal instance is hugely off from the intended distribution, and repeats the same egg-heavy drop pattern every day.  Is it possible to evaluate the actual returns on drops to see if the observed distribution across the game - even tested on an isolated day - is anything like the intent?

A lot of commentary above is on the intent of the coded drop rate as this rate will of course impact game experience and reward goals, but if the game isn't delivering that intent in practice, then it seems that producing the intended distribution would need attention first?  If my experience is an anomaly, fair enough.  But the daily repeat has eventually prompted me to at least ask the question.

**Comment by Alys** [comment]
@MaidM predictableRandom is defined here: https://github.com/HabitRPG/habitrpg/blob/develop/common/script/index.coffee#L1217-1221
Yes, it is using user.stats.gp as a seed. You should be able to find most/all of the related code in that same file.

My anecdote is that I seem to get roughly equal numbers of the three kinds of drops. I haven't noticed any long runs of one kind. I did for a while notice that I was NEVER getting red food, but then when I finally got the pieces I needed and I looked back at the other red pets I needed to feed, I realised there were none - my extended period of no red foods had been balanced out in the past by more red foods than average, but I simply hadn't noticed that at the time. Humans are usually not good at accurately analysing patterns over the long term because we tend to pay more attention to specific instances that have immediate meaning for us.

Having said that though, you are welcome to examine the code and do whatever tests you want, and if you find problems or ways to improve it, please do submit a pull request. If you aren't yet familiar with our code, you can learn about it from http://habitrpg.wikia.com/wiki/Guidance_for_Blacksmiths and related pages.

**Comment by SilkAndSlug** [comment]
I'm new to Habitica's code, and fairly new to the game and community: please forgive me if this has been said elsewhere! 

FYI, I have 28 pets and 1 mount.

Could we alter the drop-ratios depending on number of pets and/or mounts? I'm thinking that new players mostly want eggs and potions but that, as their Stable grows, they'll mostly want food. I've not run the numbers, but for an example:
- start at 50/50/0 eggs/potions/food;
- when pets reach 9/90, change the ratio to 30/30/40;
- when pets reach 30/90, change the ratio to 25/25/50;
- when pets reach 60/90, 20/20/60;
- when pets reach 90/90, 15/15/70

The aim being to _slowly_ and linearly (ish) reduce the pet/mount collection-rate from the game-start to endgame, and to avoid frustrating mid-game players by prioritising pets, then mounts+pets.

**Comment by lemoness** [comment]
Since this issue was opened and last commented on, we have released a new feature, the Enchanted Armoire, which allows you to get food drops. Furthermore, because it is based on gold, selling excess eggs and potions means that you can click on the Armoire more frequently. Additionally, we also give out multiple times a year food as part of Seasonal events, although this has a lesser impact. 

Two further points: we now release monthly pet quests (requiring normal potions to hatch), and magic hatching potions (requiring normal eggs to hatch), so players of all levels now have reasons to want eggs and potions as well as food.

Given these notable changes, I'm closing this ticket, but feel free to comment if you disagree!
